### PR TITLE
in_dummy: refactor to use config_map.

### PR DESCRIPTION
**Pull request by pwhelan**
Add configmap support for the in_dummy plugin. This is related to https://github.com/fluent/fluent-bit/issues/4863.

----
**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Review by nokute78 — CHANGES_REQUESTED** [review]
I can't accept it.
Please writing a test code before refactoring or testing to check breaking changes...

This patch breaks timestamp of output.
```
$ bin/fluent-bit -i dummy -o stdout
Fluent Bit v1.9.0
* Copyright (C) 2015-2021 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/02/26 12:36:42] [ info] [engine] started (pid=37277)
[2022/02/26 12:36:42] [ info] [storage] version=1.1.6, initializing...
[2022/02/26 12:36:42] [ info] [storage] in-memory
[2022/02/26 12:36:42] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/02/26 12:36:42] [ info] [cmetrics] version=0.3.0
[2022/02/26 12:36:42] [ info] [sp] stream processor started
[2022/02/26 12:36:42] [ info] [output:stdout:stdout.0] worker #0 started
[0] dummy.0: [1.000000000, {"message"=>"dummy"}]
[1] dummy.0: [1.999713421, {"message"=>"dummy"}]
[2] dummy.0: [2.999711601, {"message"=>"dummy"}]
[3] dummy.0: [4.000132393, {"message"=>"dummy"}]
```

> This is related to https://github.com/fluent/fluent-bit/issues/4863.

in_dummy already supports config map. I don't think it needs refactoring.

File: plugins/in_dummy/in_dummy.c
Comment (pwhelan) [reply plugins/in_dummy/in_dummy.c:171]: I tested and now have it working in [26471811d].

File: plugins/in_dummy/in_dummy.c
Comment (pwhelan) [reply plugins/in_dummy/in_dummy.c:263]: I have refactored the code to work properly by using flb_input_get_property. The code though is now extremely complex because the original code works in a very weird way where it uses the default when the parameter is invalid JSON.

```c
    /* handle it explicitly since we need to validate it is valid JSON */
    msg = flb_input_get_property("dummy", in);
    if (msg) {
        ret = flb_pack_json(msg, strlen(msg), &ctx->ref_msgpack,
                            &ctx->ref_msgpack_size, &root_type);
    }
    if (msg && ret == 0) {
    	ctx->dummy_message = flb_strdup(msg);
    	ctx->dummy_message_len = strlen(msg);
    } else  {
        flb_plg_warn(ctx->ins, "data is incomplete. Use default string.");

        ctx->dummy_message = flb_strdup(DEFAULT_DUMMY_MESSAGE);
        ctx->dummy_message_len = strlen(ctx->dummy_message);

        ret = flb_pack_json(ctx->dummy_message,
                            ctx->dummy_message_len,
                            &ctx->ref_msgpack, &ctx->ref_msgpack_size,
                            &root_type);
        if (ret != 0) {
            flb_plg_error(ctx->ins, "unexpected error");
            return -1;
        }
    }
```

I'm trying to think of a way to both support the original behavior and reduce the messy conditional complexity. Any suggestions?

Comment (pwhelan) [reply plugins/in_dummy/in_dummy.c]: I refactored it to this for now:

```c
    /* handle it explicitly since we need to validate it is valid JSON */
    msg = flb_input_get_property("dummy", in);
    if (msg == NULL) {
        msg = DEFAULT_DUMMY_MESSAGE;
    }
    ret = flb_pack_json(msg, strlen(msg), &ctx->ref_msgpack,
                        &ctx->ref_msgpack_size, &root_type);
    if (ret == 0) {
        ctx->dummy_message = flb_strdup(msg);
        ctx->dummy_message_len = strlen(msg);
    } else {
        flb_plg_warn(ctx->ins, "data is incomplete. Use default string.");

        ctx->dummy_message = flb_strdup(DEFAULT_DUMMY_MESSAGE);
        ctx->dummy_message_len = strlen(ctx->dummy_message);

        ret = flb_pack_json(ctx->dummy_message,
                            ctx->dummy_message_len,
                            &ctx->ref_msgpack, &ctx->ref_msgpack_size,
                            &root_type);
        if (ret != 0) {
            flb_plg_error(ctx->ins, "unexpected error");
            return -1;
        }
    }
```

Still not entirely satisfied with it. If DEFAULT_DUMMY_MESSAGE is, for some weird reason, invalid JSON ti will parse it twice, but that should never happen in practice and that case is still handled.

**Comment by pwhelan** [comment]
Two runs with valgrind to make sure dummy is handled correctly:

```
$ valgrind --leak-check=full ./build/bin/fluent-bit -i dummy -p start_time_sec=1337 -p start_time_nsec=1445 -o stdout -m '*' -o exit -m '*' -p flush_count=3 -f 1
==4035891== Memcheck, a memory error detector
==4035891== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==4035891== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==4035891== Command: ./build/bin/fluent-bit -i dummy -p start_time_sec=1337 -p start_time_nsec=1445 -o stdout -m * -o exit -m * -p flush_count=3 -f 1
==4035891== 
Fluent Bit v1.9.0
* Copyright (C) 2015-2021 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/02/28 10:18:04] [ info] [engine] started (pid=4035891)
[2022/02/28 10:18:04] [ info] [storage] version=1.1.6, initializing...
[2022/02/28 10:18:04] [ info] [storage] in-memory
[2022/02/28 10:18:04] [ info] [output:stdout:stdout.0] worker #0 started
[2022/02/28 10:18:04] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/02/28 10:18:04] [ info] [cmetrics] version=0.3.0
[2022/02/28 10:18:04] [ info] [sp] stream processor started
[0] dummy.0: [1337.000001445, {"message"=>"dummy"}]
[0] dummy.0: [1338.001354340, {"message"=>"dummy"}]
[0] dummy.0: [1338.991644799, {"message"=>"dummy"}]
[2022/02/28 10:18:08] [ warn] [engine] service will shutdown in max 5 seconds
[2022/02/28 10:18:09] [ info] [engine] service has stopped (0 pending tasks)
[2022/02/28 10:18:09] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/02/28 10:18:09] [ info] [output:stdout:stdout.0] thread worker #0 stopped
/in_dummy.c.gcda:overwriting an existing profile data with a different timestamp
==4035891== 
==4035891== HEAP SUMMARY:
==4035891==     in use at exit: 0 bytes in 0 blocks
==4035891==   total heap usage: 2,284 allocs, 2,284 frees, 1,673,310 bytes allocated
==4035891== 
==4035891== All heap blocks were freed -- no leaks are possible
==4035891== 
==4035891== For lists of detected and suppressed errors, rerun with: -s
==4035891== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
$ valgrind --leak-check=full ./build/bin/fluent-bit -i dummy -p dummy='{"FOO": "BAR"}' -p start_time_sec=1337 -p start_time_nsec=1445 -o stdout -m '*' -o exit -m '*' -p flush_count=3 -f 1
==4035987== Memcheck, a memory error detector
==4035987== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==4035987== Using Valgrind-3.18.1 and LibVEX; rerun with -h for copyright info
==4035987== Command: ./build/bin/fluent-bit -i dummy -p dummy={"FOO":\ "BAR"} -p start_time_sec=1337 -p start_time_nsec=1445 -o stdout -m * -o exit -m * -p flush_count=3 -f 1
==4035987== 
Fluent Bit v1.9.0
* Copyright (C) 2015-2021 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/02/28 10:18:25] [ info] [engine] started (pid=4035987)
[2022/02/28 10:18:25] [ info] [storage] version=1.1.6, initializing...
[2022/02/28 10:18:25] [ info] [output:stdout:stdout.0] worker #0 started
[2022/02/28 10:18:25] [ info] [storage] in-memory
[2022/02/28 10:18:25] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/02/28 10:18:25] [ info] [cmetrics] version=0.3.0
[2022/02/28 10:18:25] [ info] [sp] stream processor started
[0] dummy.0: [1337.000001445, {"FOO"=>"BAR"}]
[0] dummy.0: [1338.000853380, {"FOO"=>"BAR"}]
[0] dummy.0: [1338.989803552, {"FOO"=>"BAR"}]
[2022/02/28 10:18:29] [ warn] [engine] service will shutdown in max 5 seconds
[2022/02/28 10:18:30] [ info] [engine] service has stopped (0 pending tasks)
[2022/02/28 10:18:30] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/02/28 10:18:30] [ info] [output:stdout:stdout.0] thread worker #0 stopped
==4035987== 
==4035987== HEAP SUMMARY:
==4035987==     in use at exit: 0 bytes in 0 blocks
==4035987==   total heap usage: 2,291 allocs, 2,291 frees, 1,673,481 bytes allocated
==4035987== 
==4035987== All heap blocks were freed -- no leaks are possible
==4035987== 
==4035987== For lists of detected and suppressed errors, rerun with: -s
==4035987== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by pwhelan** [comment]
Debug run:

```
./build/bin/fluent-bit -vvvv -i dummy -p dummy='{"FOO": "BAR"}' -o stdout -m '*' -o exit -m '*' -p flush_count=3 -f 1
Fluent Bit v1.9.0
* Copyright (C) 2015-2021 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/02/28 10:26:08] [ info] Configuration:
[2022/02/28 10:26:08] [ info]  flush time     | 1.000000 seconds
[2022/02/28 10:26:08] [ info]  grace          | 5 seconds
[2022/02/28 10:26:08] [ info]  daemon         | 0
[2022/02/28 10:26:08] [ info] ___________
[2022/02/28 10:26:08] [ info]  inputs:
[2022/02/28 10:26:08] [ info]      dummy
[2022/02/28 10:26:08] [ info] ___________
[2022/02/28 10:26:08] [ info]  filters:
[2022/02/28 10:26:08] [ info] ___________
[2022/02/28 10:26:08] [ info]  outputs:
[2022/02/28 10:26:08] [ info]      stdout.0
[2022/02/28 10:26:08] [ info]      exit.1
[2022/02/28 10:26:08] [ info] ___________
[2022/02/28 10:26:08] [ info]  collectors:
[2022/02/28 10:26:08] [ info] [engine] started (pid=4037678)
[2022/02/28 10:26:08] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2022/02/28 10:26:08] [debug] [storage] [cio stream] new stream registered: dummy.0
[2022/02/28 10:26:08] [ info] [storage] version=1.1.6, initializing...
[2022/02/28 10:26:08] [ info] [storage] in-memory
[2022/02/28 10:26:08] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/02/28 10:26:08] [ info] [cmetrics] version=0.3.0
[2022/02/28 10:26:08] [debug] [dummy:dummy.0] created event channels: read=19 write=20
[2022/02/28 10:26:08] [debug] [stdout:stdout.0] created event channels: read=21 write=22
[2022/02/28 10:26:08] [debug] [exit:exit.1] created event channels: read=28 write=29
[2022/02/28 10:26:08] [debug] [router] match rule dummy.0:stdout.0
[2022/02/28 10:26:08] [debug] [router] match rule dummy.0:exit.1
[2022/02/28 10:26:08] [ info] [sp] stream processor started
[2022/02/28 10:26:08] [ info] [output:stdout:stdout.0] worker #0 started
[2022/02/28 10:26:09] [debug] [input chunk] update output instances with new chunk size diff=20
[2022/02/28 10:26:10] [debug] [task] created task=0x7ffff5e4a150 id=0 OK
[2022/02/28 10:26:10] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] dummy.0: [1337.000001445, {"FOO"=>"BAR"}]
[2022/02/28 10:26:10] [debug] [input chunk] update output instances with new chunk size diff=20
[2022/02/28 10:26:10] [debug] [out flush] cb_destroy coro_id=0
[2022/02/28 10:26:10] [debug] [out flush] cb_destroy coro_id=0
[2022/02/28 10:26:10] [debug] [task] destroy task=0x7ffff5e4a150 (task_id=0)
[2022/02/28 10:26:11] [debug] [task] created task=0x7ffff5e4a150 id=0 OK
[2022/02/28 10:26:11] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] dummy.0: [1338.000024463, {"FOO"=>"BAR"}]
[2022/02/28 10:26:11] [debug] [input chunk] update output instances with new chunk size diff=20
[2022/02/28 10:26:11] [debug] [out flush] cb_destroy coro_id=1
[2022/02/28 10:26:11] [debug] [out flush] cb_destroy coro_id=1
[2022/02/28 10:26:11] [debug] [task] destroy task=0x7ffff5e4a150 (task_id=0)
[2022/02/28 10:26:12] [debug] [task] created task=0x7ffff5e4a150 id=0 OK
[2022/02/28 10:26:12] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] dummy.0: [1339.000004879, {"FOO"=>"BAR"}]
[2022/02/28 10:26:12] [debug] [input chunk] dummy.0 is paused, cannot append records
[2022/02/28 10:26:12] [ warn] [engine] service will shutdown in max 5 seconds
[2022/02/28 10:26:12] [debug] [out flush] cb_destroy coro_id=2
[2022/02/28 10:26:12] [debug] [out flush] cb_destroy coro_id=2
[2022/02/28 10:26:12] [debug] [task] destroy task=0x7ffff5e4a150 (task_id=0)
[2022/02/28 10:26:13] [debug] [input chunk] dummy.0 is paused, cannot append records
[2022/02/28 10:26:13] [ info] [engine] service has stopped (0 pending tasks)
[2022/02/28 10:26:13] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/02/28 10:26:13] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

One without setting start_time_sec/start_time_nsec for good measure:

```
./build/bin/fluent-bit -vvvv -i dummy -p dummy='{"FOO": "BAR"}' -o stdout -m '*' -o exit -m '*' -p flush_count=3 -f 1
Fluent Bit v1.9.0
* Copyright (C) 2015-2021 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/02/28 10:27:19] [ info] Configuration:
[2022/02/28 10:27:19] [ info]  flush time     | 1.000000 seconds
[2022/02/28 10:27:19] [ info]  grace          | 5 seconds
[2022/02/28 10:27:19] [ info]  daemon         | 0
[2022/02/28 10:27:19] [ info] ___________
[2022/02/28 10:27:19] [ info]  inputs:
[2022/02/28 10:27:19] [ info]      dummy
[2022/02/28 10:27:19] [ info] ___________
[2022/02/28 10:27:19] [ info]  filters:
[2022/02/28 10:27:19] [ info] ___________
[2022/02/28 10:27:19] [ info]  outputs:
[2022/02/28 10:27:19] [ info]      stdout.0
[2022/02/28 10:27:19] [ info]      exit.1
[2022/02/28 10:27:19] [ info] ___________
[2022/02/28 10:27:19] [ info]  collectors:
[2022/02/28 10:27:19] [ info] [engine] started (pid=4037878)
[2022/02/28 10:27:19] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2022/02/28 10:27:19] [debug] [storage] [cio stream] new stream registered: dummy.0
[2022/02/28 10:27:19] [ info] [storage] version=1.1.6, initializing...
[2022/02/28 10:27:19] [ info] [storage] in-memory
[2022/02/28 10:27:19] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/02/28 10:27:19] [ info] [cmetrics] version=0.3.0
[2022/02/28 10:27:19] [debug] [dummy:dummy.0] created event channels: read=19 write=20
[2022/02/28 10:27:19] [debug] [stdout:stdout.0] created event channels: read=21 write=22
[2022/02/28 10:27:19] [debug] [exit:exit.1] created event channels: read=28 write=29
[2022/02/28 10:27:19] [debug] [router] match rule dummy.0:stdout.0
[2022/02/28 10:27:19] [debug] [router] match rule dummy.0:exit.1
[2022/02/28 10:27:19] [ info] [sp] stream processor started
[2022/02/28 10:27:19] [ info] [output:stdout:stdout.0] worker #0 started
[2022/02/28 10:27:19] [debug] [input chunk] update output instances with new chunk size diff=20
[2022/02/28 10:27:20] [debug] [task] created task=0x7ffff5e4a150 id=0 OK
[2022/02/28 10:27:20] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2022/02/28 10:27:20] [debug] [input chunk] update output instances with new chunk size diff=20
[0] dummy.0: [1646054839.097209334, {"FOO"=>"BAR"}]
[2022/02/28 10:27:20] [debug] [out flush] cb_destroy coro_id=0
[2022/02/28 10:27:20] [debug] [out flush] cb_destroy coro_id=0
[2022/02/28 10:27:20] [debug] [task] destroy task=0x7ffff5e4a150 (task_id=0)
[2022/02/28 10:27:21] [debug] [task] created task=0x7ffff5e4a150 id=0 OK
[2022/02/28 10:27:21] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] dummy.0: [1646054840.097240215, {"FOO"=>"BAR"}]
[2022/02/28 10:27:21] [debug] [out flush] cb_destroy coro_id=1
[2022/02/28 10:27:21] [debug] [input chunk] update output instances with new chunk size diff=20
[2022/02/28 10:27:21] [debug] [out flush] cb_destroy coro_id=1
[2022/02/28 10:27:21] [debug] [task] destroy task=0x7ffff5e4a150 (task_id=0)
[2022/02/28 10:27:22] [debug] [task] created task=0x7ffff5e4a150 id=0 OK
[2022/02/28 10:27:22] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2022/02/28 10:27:22] [debug] [input chunk] dummy.0 is paused, cannot append records
[0] dummy.0: [1646054841.097226726, {"FOO"=>"BAR"}]
[2022/02/28 10:27:22] [ warn] [engine] service will shutdown in max 5 seconds
[2022/02/28 10:27:22] [debug] [out flush] cb_destroy coro_id=2
[2022/02/28 10:27:22] [debug] [out flush] cb_destroy coro_id=2
[2022/02/28 10:27:22] [debug] [task] destroy task=0x7ffff5e4a150 (task_id=0)
[2022/02/28 10:27:23] [debug] [input chunk] dummy.0 is paused, cannot append records
[2022/02/28 10:27:23] [ info] [engine] service has stopped (0 pending tasks)
[2022/02/28 10:27:23] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/02/28 10:27:23] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

**Comment by edsiper** [comment]
waiting for final review from @nokute78 on this one

**Comment by pwhelan** [comment]
> waiting for final review from @nokute78 on this one

It  should be ready for final by @nokute78. I have tested and fixed the problems with the timestamp and start_time_sec/start_time_nsec. I also added a unit test but the actual records being generated are not being validated.

File: plugins/in_dummy/in_dummy.c
Comment (pwhelan) [reply plugins/in_dummy/in_dummy.c]: I'm going to check that I am not double freeing the dummy_message then marked this as resolved.

**Comment by pwhelan** [comment]
@nokute78 I have added a unit test that tests the records generated for the interval_sec and interval_nsec parameters.

**Comment by edsiper** [comment]
@nokute78 can you take a look again at this pls ?

File: plugins/in_dummy/in_dummy.c
Comment (nokute78) [inline plugins/in_dummy/in_dummy.c:144]: Could you add `ctx->dummy_message = NULL` ?
In the future, the function may return before checking `dummy` property.

**Review by nokute78 — CHANGES_REQUESTED** [review]
@pwhelan Thank you for updating.
Could you add testcases?

1. No option test to detect such regression https://github.com/fluent/fluent-bit/pull/4924#pullrequestreview-894362112
1. "dummy" input test like `{"new_key":"new_value"}`
1. Invalid "dummy" input like `{"test":"message`

The case 1 should check the timestamp is valid (like saving the timestamp before emitting and comparing) and check if the record is the default message.


Could you check below case ?
https://github.com/fluent/fluent-bit/pull/4924/files#diff-259ac31effef0c86a87225f5ad013df83343367ddefb9f9c7438ef62420d676fR202
We should free `ctx->dummy_message` and filling NULL.
`ctx->dummy_timestamp` also should be freed if it is not NULL.

File: plugins/in_dummy/in_dummy.c
Comment (pwhelan) [reply plugins/in_dummy/in_dummy.c:144]: done.

**Comment by pwhelan** [comment]
> @pwhelan Thank you for updating. Could you add testcases?
> 
>     1. No option test to detect such regression [#4924 (review)](https://github.com/fluent/fluent-bit/pull/4924#pullrequestreview-894362112)
>     2. "dummy" input test like `{"new_key":"new_value"}`
>     3. Invalid "dummy" input like `{"test":"message`

I have added a test for all three cases, as well as a test against the default message.

**Pull request merged**
